### PR TITLE
New operators

### PR DESCRIPTION
**Pull request by bench1ps**
* I added two operators that may be useful : `EqualsTo`, that asserts the equality of two values, and `MatchesRegex`, that matches an argument against a regex (helpul to activate a toggle for users with some email address domain for instance). I also refactored a bit the operators to avoid code duplication.

* I feel that, in the same way serializers are used to handle operators, the context should be easily deserialized (which would be very useful in the API to check if a toggle is active). This would allow to implement a new method in the API using a POST call I guess. Note that a context should always be passed with the request, as it is mandatory in the `ToggleManager`.

Thank you for your comments.

### CODE REVIEW

File: test/Qandidate/Toggle/Operator/EqualsToOperatorTest.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    public function it_does_not_apply_to_not_equal_values($value, $argument)
+    {
+        $operator = new EqualsTo($value);
+        $this->assertFalse($operator->appliesTo($argument));
+    }
+
+    public function notEqualValues()
+    {
+        return array(
+            array(42,   43),
```
Comment (wjzijderveld) [inline test/Qandidate/Toggle/Operator/EqualsToOperatorTest.php]: We probably need some tests for edge cases' like false vs 0 etc.

File: lib/Qandidate/Toggle/Serializer/ContextSerializer.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+        $context = new Context();
+        foreach ($data['context'] as $key => $value) {
+            if (!$context->has($key)) {
+                $context->set($key, $value);
+            }
+        }
+
+        return $context;
+    }
```
Comment (wjzijderveld) [inline lib/Qandidate/Toggle/Serializer/ContextSerializer.php]: Is there a specific reason that you didn't add the serialize method?

File: lib/Qandidate/Toggle/Serializer/ContextSerializer.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+class ContextSerializer
+{
+    /**
+     * @param array $data
+     *
+     * @return Context
+     */
+    public function deserialize(array $data)
+    {
+        $this->assertHasKey('context', $data);
```
Comment (wjzijderveld) [inline lib/Qandidate/Toggle/Serializer/ContextSerializer.php]: Why shouldn't $data just be the context? Would simplify this method a lot.

**Comment by wjzijderveld** [comment]
Thanks for the PR! Added a few comments/questions..

File: lib/Qandidate/Toggle/Serializer/ContextSerializer.php (hunk 3 above)
Comment (bench1ps) [inline lib/Qandidate/Toggle/Serializer/ContextSerializer.php]: Well, the idea was to deserialize the whole request when asking for a toggle, and the request contains two entries : the name of the toggle and the context.

To make things right, using the pattern already in place for the toggle adding request for instance, I'm guessing we should have another serializer that would have the responsibility to check for a 'name' and 'context' entries and then call the `ContextSerializer` when needed, but it kind of seems a bit overkill to me.

What do you think ? Thank you for your feedback BTW :)

File: lib/Qandidate/Toggle/Serializer/ContextSerializer.php (hunk 2 above)
Comment (bench1ps) [inline lib/Qandidate/Toggle/Serializer/ContextSerializer.php]: As I see this, the serializers are only useful when using the lib through the API, and a `serialize` method would only make sense if the API itself would send back the context in the response. But I couldn't think of any case where it might be needed.

File: test/Qandidate/Toggle/Operator/EqualsToOperatorTest.php (hunk 1 above)
Comment (bench1ps) [inline test/Qandidate/Toggle/Operator/EqualsToOperatorTest.php]: Thanks, I'll add some more tests :)

Comment (bench1ps) [inline test/Qandidate/Toggle/Operator/EqualsToOperatorTest.php]: I added these tests :

    array(false, 0),
    array(null, 0),
    array(true, 1),
    array("0", 0),

If you think of anything else, you let me know.

File: lib/Qandidate/Toggle/Operator/EqualsTo.php [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Qandidate\Toggle\Operator;
+
+class EqualsTo extends EqualityOperator
```
Comment (asm89) [inline lib/Qandidate/Toggle/Operator/EqualsTo.php]: `EqualTo` ?

File: lib/Qandidate/Toggle/Serializer/ContextSerializer.php (hunk 3 above)
Comment (asm89) [inline lib/Qandidate/Toggle/Serializer/ContextSerializer.php]: Do we need this class? This seems like logic for a service/controller in the API to me.

Comment (bench1ps) [inline lib/Qandidate/Toggle/Serializer/ContextSerializer.php]: I get your point, and I feel the same way for every serializer in the lib. But the truth is they are not actual serializers to me, they merely provide some useful method to flatten or reconstruct an object. But you're right, it's the control layer that should be responsible of such a behaviour, namely the API. Yet, I don't suppose you plan to remove all the serializers from the lib to make them available only in the API.

**Comment by wjzijderveld** [comment]
@bench1ps Sorry for the lack of activity from my side.

Could you rename `EqualsTo` to `EqualTo` and remove the ContextSerializer part?
You can open a new PR about the context serializer, but I don't really want to merge it at this moment. 
But the operators are nice on their own :-)

**Comment by wjzijderveld** [comment]
@bench1ps Any update here?

**Comment by asm89** [comment]
As far as I understand this was merged in #25 @wjzijderveld? If not, feel free to re-open. :)
